### DISCUSSION TITLE
How to call with a particular entity data type?

### DISCUSSION THREAD

**Discussion by Lex-Kent** [speechly/speechly#253] [General]
Hi all,

I'd like to call Speechly using the built-in entity data types: https://docs.speechly.com/reference/entity-data-types/#number

At various places in my app, I will need numbers, dates and names - which I was very pleased to see are already included in Speechly. However, I have no idea how to do this!

I am using a `Speechly BrowserClient`.

the important part of my code is:

```
ctx.speechly = new BrowserClient({
  appId: ctx.speechlyAppId,
  // debug: true,
  logSegments: false,
  vad: {
    enabled: true,
    noiseGateDb: -24.0
  }
});

ctx.speechlyMic = new BrowserMicrophone();

ctx.speechly.onSegmentChange(segment => {
  console.log(segment.words.map(w => w.value.toLowerCase()).join(" "));

  // Clean up and concatenate words
  const transcript = segment.words.map(w => w.value.toLowerCase()).join(" ");
  if (this.continuousReply){
    if (isNil(this.openBub)){
      this.openBub = openBubble(capitalise(transcript, true), 0, true);
    }
    else{
      addToBubble(this.openBub, capitalise(transcript, true));
    }
  }

  if (segment.isFinal){
    ctx.listenResult(ctx.listenE, transcript);
  }

});
```

I've searched here in the forum and in the Speechly docs, but I can't find anything to do with this. Apologies if it's been staring me in the face all day!

Many thanks,

**Comment by thisismatu** [comment]
Hi @Lex-Kent 👋 

Good news, the `segment` object actually contains all necessary info. If you log it to console, you'll find that it contains the following properties:

- `id` _(number)_ The index (zero-based) of this segment within the audio context. An audio context can consist of several consecutive segments.
- `contextId` _(string)_ - The audio context to which this segment belongs to (UUID).
- `isFinal` _(boolean)_ - A boolean that indicates if this is the last time callback is called with this segment. Subsequent calls to callback within the same audio context refer to the next segment. Note that none of the data associated with this segment will no longer be attached to the next segment.
- `intent` _(SpeechIntent)_  - The intent associated with this segment. There can only be one intent for a segment.
- `entities` _(Entity[])_ - A list of entities. There can be several entities that belong to the same segment.
- `words` _(Transcript[])_ - A list of Transcript objects. Together these contain the text produced by speech recognition.

<details>
<summary>View example output</summary>

```
{
  "id": 0,
  "contextId": "aa901781-0961-4f15-b992-5cb76eae3672",
  "isFinal": true,
  "words": [
    {
      "value": "schedule",
      "index": 2,
      "startTimestamp": 1500,
      "endTimestamp": 2160,
      "isFinal": true
    },
    {
      "value": "for",
      "index": 3,
      "startTimestamp": 2160,
      "endTimestamp": 2640,
      "isFinal": true
    },
    {
      "value": "october",
      "index": 4,
      "startTimestamp": 2640,
      "endTimestamp": 3300,
      "isFinal": true
    },
    {
      "value": "tenth",
      "index": 5,
      "startTimestamp": 3300,
      "endTimestamp": 3720,
      "isFinal": true
    }
  ],
  "entities": [
    {
      "type": "date",
      "value": "2023-10-10",
      "startPosition": 4,
      "endPosition": 6,
      "isFinal": true
    }
  ],
  "intent": {
    "intent": "propose_meeting",
    "isFinal": true
  }
```
</details>

From your code example i can see that you are using only the `words` property, which contains a list of the words in the transcript and this is verbatim. If you look inside the `entities` property you'll find a list of the entities that contain the type, post-processed value and start/end position indexes. This is probably what you want and you can use these indexes to substitute a word.

Our v1.0 docs contains a react guide on how to deal with entities and intents. Although it looks like you are using vanilla JavaScript, the same concepts apply https://dreamy-cori-a02de1.netlify.app/client-libraries/using-react-client/#intents-and-entities

Happy developing!

**Reply by Lex-Kent** [reply]
@thisismatu thanks so much for taking the time to help.

I now understand that my segments contain other properties aside from just words, but I still don't see how I submit my request with an intent before getting back the segments from the API.

I'm beginning to suspect I need to create training data first, and include the intents and entities in the training data, is that correct? If so, that's a real shame - it would be incredibly useful to submit any dialogue to Speechly with an expectation of what the user should be saying - an age if they've been asked their age, an email address if they've been asked for that, a future date if they've been asked for one etc. etc. without the need to create training data.

**Reply by thisismatu** [reply]
Yes, that is correct. Out of the box speechly operates in text-to-speech mode, meaning it will only produce a verbatim transcript. If you want it pick up entities and intents and do some downstream actions based on them, you need to train the model it with training data. For your case i would go ahead and define some intents and entities along with annotated example utterances. Once there's enough data, the system will be able to generalize. We have a whole section on intents and entity detection in our docs https://docs.speechly.com/features/intents-entities

**Comment by cfry** [comment]
Is there a way to configure a Speechly App such that all utterances will have all the numbers in them converted to number entities?

**Reply by teelisyys** [reply]
Support for converting all numbers to entities is not a generally available feature currently.

Also, we are working right now to get a more straight forward support for digits for the _transcription_ use case. For some downstream tasks this might bypass the need for number _entities_.

Depending on the concrete use case and downstream task, for our Enterprise plan customers we might have available solutions which are not yet publicly documented. [You can learn more about our different plans on our website](https://www.speechly.com/pricing).

**Comment by cfry** [comment]
Great! I can imagine numbers being a special sort of entity that
doesn't use the regular mechanism for other entities in Speechly.

There's at least 2 reasons why they perhaps are best special-cased:
- they're so commonly needed
- though there's an unlimited set of them, we can clearly recognize them.
